gltf-importer/src/lib.rs

```diff
@@ -241,73 +152,58 @@ fn load_external_buffers(
     use config::ValidationStrategy;
-    use gltf::json::validation::Validate;
-    match config.validation_strategy {
-        ValidationStrategy::Skip => Ok(()),
-        ValidationStrategy::Minimal => {
-            let mut errs = vec![];
-            json.validate_minimally(
-                &json,
-                || json::Path::new(),
-                &mut |path, err| errs.push((path(), err)),
-            );
-            if errs.is_empty() {
-                Ok(())
-            } else {
-                Err(Error::Validation(errs))
-            }
-        },
-        ValidationStrategy::Complete => {
-            let mut errs = vec![];
-            json.validate_completely(
-                &json,
-                || json::Path::new(),
-                &mut |path, err| errs.push((path(), err)),
-            );
-            if errs.is_empty() {
-                Ok(())
-            } else {
-                Err(Error::Validation(errs))
-            }
-        },
-    }
+    Ok(match config.validation_strategy {
+        ValidationStrategy::Skip => unsafe { unvalidated.skip_validation() },
+        ValidationStrategy::Minimal => unvalidated.validate_minimally()?,
+        ValidationStrategy::Complete => unvalidated.validate_completely()?,
+    })
 }
 
 fn validate_binary(
-    json: &gltf::json::Root,
+    unvalidated: gltf::Unvalidated,
     config: &Config,
-    has_blob: bool,
-) -> Result<(), Error> {
-    use self::json::validation::Error as Reason;
-    let mut errs = vec![];
-    let _ = validate_standard(json, config)?;
-    if config.validation_strategy == config::ValidationStrategy::Skip {
-        return Ok(());
+    has_bin: bool,
+) -> Result<Gltf, Error> {
+    use config::ValidationStrategy;
+    use json::validation::Error as Reason;
+
+    if config.validation_strategy == ValidationStrategy::Skip {
+        return Ok(unsafe { unvalidated.skip_validation() });
     }
 
-    // Required for the `Minimal` and `Complete` validation strategies.
-    for (index, buffer) in json.buffers.iter().enumerate() {
-        let path = || {
-            json::Path::new()
-                .field("buffers")
-                .index(index)
-                .field("uri")
-        };
-        match index {
-            0 if has_blob => if buffer.uri.is_some() {
-                errs.push((path(), Reason::Missing));
-            },
-            _ if buffer.uri.is_none() => {
-                errs.push((path(), Reason::Missing));
-            },
-            _ => {},
+    let mut errs = vec![];
+    {
+        let json = unvalidated.as_json();
+        for (index, buffer) in json.buffers.iter().enumerate() {
+            let path = || {
+                json::Path::new()
+                    .field("buffers")
+                    .index(index)
+                    .field("uri")
+            };
+            match index {
+                0 if has_bin => if buffer.uri.is_some() {
+                    errs.push((path(), Reason::Missing));
```

Comment:
The glTF JSON slightly different rules for buffers depending on whether it comes from a .gltf file or a .glb file. The former requires URIs for all buffers, whereas the latter requires the first (and only the first) to have no URI if and only if a BIN chunk is provided.

Currently the buffer URI isn't exposed to user although it needs to be. I'm deciding between two ideas: a `Data` enum with `Bin` and `Uri` variants, or returning a special URI (e.g. `"#bin"`) to denote an embedded buffer.